### PR TITLE
MediaPlugin

### PR DESCRIPTION
**Pull request by vogdb**

### CODE REVIEW

**Comment by vogdb** [comment]
@tomaszdurka or @zazabe please review

**Comment by vogdb** [comment]
@zazabe please review

**Comment by zazabe** [comment]
- some missing docus
- not sure if we really need `IceCandidateListener` / `IceCandidate`, is it not possible to just deal with plain object?  
- https://www.npmjs.com/package/webrtc-adapter is not supposed to provide the shim for `getUserMedia`, etc... ?
- the `pc.onicecandidate` callback should be registered when the `PeerConnection` is instantiated, because we may listen multiple times to `candidate/complete` in `MediaPlugin::_createSDP`  if an offer is created first, and then an answer
- not sure, but in `janus.js`, for `createOffer/createAnswer`, all candidates must have been received before calling the success callback: https://github.com/meetecho/janus-gateway/blob/master/html/janus.js#L1641    
  I don't understand exactly what will happen if `(!config.iceDone && !config.trickle) == true`, obviously, no `callbacks.success`, but then... ?

**Comment by vogdb** [comment]
> - some missing docus

ok
> - https://www.npmjs.com/package/webrtc-adapter is not supposed to provide the shim for `getUserMedia`, etc... ?

It does not provide `screen` constraint for `getUserMedia` that janus supports. I didn't find any alternative to `screen` also in specs: http://w3c.github.io/mediacapture-main/#dom-mediadevices-getusermedia so in order to keep it, I took its janus implementation.

> - not sure if we really need `IceCandidateListener` / `IceCandidate`, is it not possible to just deal with plain object?
- the `pc.onicecandidate` callback should be registered when the `PeerConnection` is instantiated, because we may listen multiple times to `candidate/complete` in `MediaPlugin::_createSDP`  if an offer is created first, and then an answer
 - not sure, but in `janus.js`, for `createOffer/createAnswer`, all candidates must have been received before calling the success callback: https://github.com/meetecho/janus-gateway/blob/master/html/janus.js#L1641    
  I don't understand exactly what will happen if `(!config.iceDone && !config.trickle) == true`, obviously, no `callbacks.success`, but then... ?

All those are connected. The place that you have pointed https://github.com/meetecho/janus-gateway/blob/master/html/janus.js#L1641 shows how `trickle` option works. When `trickle` is true, it calls `success` callback as soon as it has SDP. When `trickle` is false, it waits for all ICE candidates before `success` callback. By default `trickle` is true. And that is the only option that is currently implemented because it is a default behaviour of SDP exchange according to https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Signaling_and_video_calling#Signaling_transaction_flow. We can implement the second option `trickle == false`. In that case `IceCandidateListener/IceCandidate` will be very helpful to handle this. If we decide that the only `trickle == true` is enough for us then we can easily remove `IceCandidateListener/IceCandidate`.

**Comment by vogdb** [comment]
@zazabe please look

**Comment by vogdb** [comment]
- user `webrtc` in Shim for getting `getUserMedia`
 - shim `PeerConnection.getStats`

**Comment by vogdb** [comment]
>  - user `webrtc` in Shim for getting `getUserMedia`

Can't do that cause `webrtc-adapter` does not export shimmed `getUserMedia`. It modifies the global `navigator.mediaDevices` :( So, we need just to require `webrtc-adapter` before scripts that depend on its shims. Please look https://github.com/webrtc/adapter/blob/master/src/js/adapter_core.js#L46 and https://github.com/webrtc/adapter/blob/master/src/js/chrome/getusermedia.js#L13

>  - shim `PeerConnection.getStats`

The same as above.

@zazabe please review

File: src/webrtc/media-devices-shim.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+              stream.onended();
+            }
+          }
+          lastStreamTime = stream.currentTime;
+        }, 500);
+      }
+      return stream;
+    });
+  } else {
+    //var error = new Error('NavigatorUserMediaError');
```
Comment (zazabe) [inline src/webrtc/media-devices-shim.js]: should it be removed?

**Comment by zazabe** [comment]
> So, we need just to require webrtc-adapter before scripts 

OK, about the browser version, I suppose that these shims will be applied automatically if we use the "browserified" version of the `janus-gateway-js`, if yes, i think it's fine.

**Comment by vogdb** [comment]
> OK, about the browser version, I suppose that these shims will be applied automatically if we use the "browserified" version of the `janus-gateway-js`, if yes, i think it's fine.

Yes. They will be applied automatically.

**Comment by vogdb** [comment]
@zazabe please review

**Comment by zazabe** [comment]
as discussed, we should listen to `RTCPeerConnection.iceConnectionState` changes and behave according to this state, but let's do it in a new PR.

lgtm!

**Comment by vogdb** [comment]
thank you!

**Pull request merged**